tools/wake-format/emitter.h

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <wcl/optional.h>
+#include <wcl/rope.h>
+
+#include <cassert>
+#include <iostream>
+#include <vector>
+
+#include "parser/cst.h"
+
+class Emitter {
+ public:
+  struct ctx_t {
+    int nest_level = 0;
+    bool is_flat = false;
+
+    ctx_t flat() {
```

Comment:
Maybe instead of having a `nest_t` we instead just make an `nest()` method like `flat()`, that way its completely functional and we don't need nest_t. That said maybe you want to use `ctx` almost everywhere and temporarily mutating a ctx is a nicer interface. Up to you.